content/program-testing.md

```diff
@@ -0,0 +1,534 @@
+let (comment_pda, _bump_seed) =
+    Pubkey::find_program_address(&[review_pda.as_ref(), b"comment"], &program_id);
+
+// Create user associate token account of token mint
+let init_ata_ix: Instruction = create_associated_token_account(
+    &payer.pubkey(),
+    &payer.pubkey(),
+    &mint_keypair.pubkey(),
+);
+
+let user_ata: Pubkey =
+    get_associated_token_address(&payer.pubkey(), &mint_keypair.pubkey());
+```
+
+Once we have all of the accounts initialized, we can put it all together into a single transaction.
+
+```rust
+// Concat data to single buffer
+let mut data_vec = vec![0];
+data_vec.append(
+    &mut (TryInto::<u32>::try_into(title.len()).unwrap().to_le_bytes())
+        .try_into()
+        .unwrap(),
+);
+data_vec.append(&mut title.into_bytes());
+data_vec.push(RATING);
+data_vec.append(
+    &mut (TryInto::<u32>::try_into(review.len())
+        .unwrap()
+        .to_le_bytes())
+    .try_into()
+    .unwrap(),
+);
+data_vec.append(&mut review.into_bytes());
+
+// Create transaction object with instructions, accounts, and input data
+let mut transaction = Transaction::new_with_payer(
+    &[
+      create_mint_acct_ix,
+      init_mint_ix,
+      init_ata_ix,
+      Instruction {
+          program_id: program_id,
+          accounts: vec![
+              AccountMeta::new_readonly(payer.pubkey(), true),
+              AccountMeta::new(review_pda, false),
+              AccountMeta::new(comment_pda, false),
+              AccountMeta::new(mint_keypair.pubkey(), false),
+              AccountMeta::new_readonly(mint_auth, false),
+              AccountMeta::new(user_ata, false),
+              AccountMeta::new_readonly(SYSTEM_PROGRAM_ID, false),
+              AccountMeta::new_readonly(TOKEN_PROGRAM_ID, false),
+          ],
+          data: data_vec,
+      },
+    ],
+    Some(&payer.pubkey()),
+);
+transaction.sign(&[&payer, &mint_keypair], recent_blockhash);
+
+// Process transaction and compare the result
+assert_matches!(banks_client.process_transaction(transaction).await, Ok(_));
+```
+
+You can now run this test with `cargo test-bpf` and if it’s successful, you’ll be able to see the program logs and the final test result in the terminal. It may take a while to compile and run the test. Feel free to take a look at the solution code here.
+
+Take a look at the testing script in the `ts` directory and compare it to what we just wrote in Rust. They are doing almost the exact same thing, but seeing how the code differs between Typescript and Rust can be eye-opening sometimes.
+
+# Challenge
+
+We just wrote a single unit test in Rust, but a proper testing architecture is made up of more than just one test. As a challenge, build on top of what we just did and write some unit tests in Rust (or Typescript if you've seen enough Rust for the day) that test the other instructions in the program and also think about how you can write some tests with malicious or inaccurate code that's supposed to return an error from the program.
```

Comment:
@ixmorrow please make this at least two sentences :)  Maybe just break it in the middle after the parentheses?